modules/network/src/main/java/org/apache/ignite/internal/network/serialization/marshal/MarshallingValidations.java

```diff
@@ -19,14 +19,25 @@
 
 import java.lang.reflect.Field;
 import java.lang.reflect.Modifier;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
 import org.apache.ignite.internal.network.serialization.Classes;
 import org.jetbrains.annotations.Nullable;
 
 /**
  * Validations that are run before marshalling objects.
  */
 class MarshallingValidations {
-    static void throwIfMarshallingNotSupported(@Nullable Object object) {
+    private final Map<Class<?>, YesNo> whetherInnerClasses = new HashMap<>();
+    private final Map<Class<?>, YesNo> whetherCapturingClosures = new HashMap<>();
+    private final Map<Class<?>, YesNo> whetherNonSerializableLambdas = new HashMap<>();
+
+    private final Function<Class<?>, YesNo> isInnerClassFunction = MarshallingValidations.this::isInnerClass;
+    private final Function<Class<?>, YesNo> isCapturingClosureFunction = this::isCapturingClosure;
+    private final Function<Class<?>, YesNo> isNonSerializableLambdaFunction = this::isNonSerializableLambda;
+
+    void throwIfMarshallingNotSupported(@Nullable Object object) {
```

Comment:
I think we should move these checks to the descriptor generation phase and generate "invalid" descriptors if we can't marshall the class

Reply:
As agreed on a call, left the caching approach, but squashed the maps. The idea is that, if we use a descriptor as a medium of the knowledge about 'marshallability' of a class, we still need to look it up in a map (a different map), but this requires to make quite some changes, and it does not seem to be worth it.